File: retrocpu_boot_monitor/test/mn1613/handshake/handshake_undef_test.ts
/**
 * g_bios_undef_led（CPU→IO コマンド 13h）
 * 根拠: HandShake.mdc「未定義命令実行通知」/ boot_monitor.mdc / test_framework.mdc
 */
import {
  createSessionFromSettings,
  expect,
  test,
  type IoBoardHandshakeMock,
  type Mn1613AsmSession,
} from "../../../../retrocpu_test_framework_ts/src/index.js";
import {
  mn1613MonHandshakeSettings,
  withMn1613CpuLog,
} from "../mn1613_mon_settings.js";

const BASE_REGS = {
  R2: 0x2222,
  R3: 0x3333,
  R4: 0x4444,
} as const;

const session: Mn1613AsmSession = createSessionFromSettings(
  withMn1613CpuLog(mn1613MonHandshakeSettings, import.meta.url),
);

/**
 * g_main 済み＋ ioMock handshake で 1 ケースを実行する。
 * @param fn 本体（session / mock 利用可）
 */
async function withCase(
  fn: (s: Mn1613AsmSession, mock: IoBoardHandshakeMock) => Promise<void>,
): Promise<void> {
  session.reload();
  try {
    await session.runInit();
    await fn(session, session.requireHandshakeMock());
  } finally {
    await session.detachIoMock();
  }
}

/**
 * g_bios_undef_led を呼び、CPU→IO 1 トランザクションと並行する。
 * @param mock IO モック
 * @param on 互換引数（旧 Bit0 指定。現仕様では未使用）
 */
async function callUndefLed(
  mock: IoBoardHandshakeMock,
  on: number,
): Promise<void> {
  const io = mock.handleOneRequest();
  await session.call("g_bios_undef_led", {
    registers: { ...BASE_REGS, R0: on },
  });
  await io;
}

test("点灯(1)で IO の undefLed が true になる", async () => {
  await withCase(async (s, mock) => {
    expect(mock.state.undefLed).toBe(false);
    await callUndefLed(mock, 1);
    s.expectRegisters({ R0: 0 });
    expect(mock.state.undefLed).toBe(true);
    expect(mock.state.lastUndefNotify !== null).toBe(true);
  });
});

test("引数 0 でも 13h 通知として処理され、undefLed は true になる", async () => {
  await withCase(async (s, mock) => {
    expect(mock.state.undefLed).toBe(false);
    await callUndefLed(mock, 0);
    s.expectRegisters({ R0: 0 });
    expect(mock.state.undefLed).toBe(true);
    expect(mock.state.lastUndefNotify !== null).toBe(true);
  });
});

test("引数 0x03 でも 13h 通知として処理される", async () => {
  await withCase(async (s, mock) => {
    await callUndefLed(mock, 0x03);
    s.expectRegisters({ R0: 0 });
    expect(mock.state.undefLed).toBe(true);
    expect(mock.state.lastUndefNotify !== null).toBe(true);
  });
});

test("R3/R4 は呼び出しの前後で保たれる", async () => {
  await withCase(async (s, mock) => {
    await callUndefLed(mock, 1);
    s.expectRegisters({ R0: 0, R3: BASE_REGS.R3, R4: BASE_REGS.R4 });
  });
});

File: retrocpu_boot_monitor/test/mn1613/handshake/handshake_undef_test.rs
use std::sync::Arc;

use retrocpu_test_framework_rs::{
    create_session_from_settings, CallOptions, CallRegisters, CodeTestIoMockEntry, FrameworkError,
    IoBoardHandshakeMock, JsonTestSettings, Mn1613AsmSession,
};

fn base_regs() -> CallRegisters {
    CallRegisters {
        r2: Some(0x2222),
        r3: Some(0x3333),
        r4: Some(0x4444),
        ..Default::default()
    }
}

fn handshake_settings() -> JsonTestSettings {
    let mut s = super::mn1613_rs_settings();
    s.io_mock = Some(vec![CodeTestIoMockEntry::Handshake]);
    s
}

fn with_case<F>(f: F) -> Result<(), FrameworkError>
where
    F: FnOnce(&mut Mn1613AsmSession, Arc<IoBoardHandshakeMock>) -> Result<(), FrameworkError>,
{
    let mut s = create_session_from_settings(&handshake_settings(), None)?;
    s.run_init()?;
    let mock = s.require_handshake_mock()?;
    let ret = f(&mut s, Arc::clone(&mock));
    s.detach_io_mock();
    ret
}

fn call_undef_led(
    s: &mut Mn1613AsmSession,
    mock: &Arc<IoBoardHandshakeMock>,
    arg: u16,
) -> Result<(), FrameworkError> {
    mock.run_with_cpu_to_io_request(|| {
        let _ = s.call(
            "g_bios_undef_led",
            CallOptions {
                registers: Some(CallRegisters {
                    r0: Some(arg),
                    ..base_regs()
                }),
                ..Default::default()
            },
        )?;
        Ok(())
    })?;
    Ok(())
}

#[test]
fn undef_notify_arg_1_returns_ok() -> Result<(), FrameworkError> {
    with_case(|s, mock| {
        assert!(!mock.undef_led());
        call_undef_led(s, &mock, 1)?;
        assert!(mock.undef_led());
        s.expect_registers(
            &CallRegisters {
                r0: Some(0),
                ..Default::default()
            },
            None,
        )
    })
}

#[test]
fn undef_notify_arg_0_returns_ok() -> Result<(), FrameworkError> {
    with_case(|s, mock| {
        assert!(!mock.undef_led());
        call_undef_led(s, &mock, 0)?;
        assert!(mock.undef_led());
        s.expect_registers(
            &CallRegisters {
                r0: Some(0),
                ..Default::default()
            },
            None,
        )
    })
}

#[test]
fn undef_notify_arg_3_returns_ok() -> Result<(), FrameworkError> {
    with_case(|s, mock| {
        call_undef_led(s, &mock, 3)?;
        assert!(mock.undef_led());
        s.expect_registers(
            &CallRegisters {
                r0: Some(0),
                ..Default::default()
            },
            None,
        )
    })
}

#[test]
fn undef_notify_preserves_r3_r4() -> Result<(), FrameworkError> {
    with_case(|s, mock| {
        call_undef_led(s, &mock, 1)?;
        s.expect_registers(
            &CallRegisters {
                r0: Some(0),
                r3: Some(0x3333),
                r4: Some(0x4444),
                ..Default::default()
            },
            None,
        )
    })
}
